src/forcing/ForcingsEngineLumpedDataProvider.cpp

```diff
@@ -40,29 +42,44 @@ Provider::ForcingsEngineLumpedDataProvider(
         };
     }
 
+    // Initialize the value cache
     const auto id_dim   = static_cast<std::size_t>(bmi_->GetVarNbytes("CAT-ID") / bmi_->GetVarItemsize("CAT-ID"));
     const auto var_dim  = get_available_variable_names().size();
 
     bmi_->Update();
     this->increment_time();
     
+    // Copy CAT-ID values into instance vector
     const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
-    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    for (int i = 0; i < id_dim; ++i) {
+        const auto divide_id = ptr[i];
+        var_divides_[divide_id] = i;
+    }
+
     var_cache_ = decltype(var_cache_){{ 2, id_dim, var_dim }};
 
     // Cache initial iteration
     update_value_storage_();
     this->increment_time(); // TODO: why???
 }
 
-Provider::ForcingsEngineLumpedDataProvider(
+BaseProvider* Provider::lumped_instance(
     const std::string& init,
     const std::string& time_begin,
     const std::string& time_end,
     const std::string& time_fmt
 )
-  : Provider(init, parse_time(time_begin, time_fmt), parse_time(time_end, time_fmt))
-{}
+{
+    auto time_begin_epoch = static_cast<size_t>(parse_time(time_begin, time_fmt));
+    auto time_end_epoch = static_cast<size_t>(parse_time(time_end, time_fmt));
+
+
+    auto provider = std::unique_ptr<ForcingsEngineLumpedDataProvider>{
+        new Provider{init, time_begin_epoch, time_end_epoch}
+    };
```

Comment:
```suggestion
    auto provider = std::make_unique<ForcingsEngineLumpedDataProvider>(
        init, time_begin_epoch, time_end_epoch
    );
```
In the `unique_ptr` case, it doesn't make a tangible difference, but it's generally good practice to use `make_unique` and `make_shared` to avoid edge conditions around exceptions, and to get better memory allocation behavior and layout in the `shared_ptr` case.

Reply:
This one can't be `std::make_unique` because the constructor for `ForcingsEngineLumpedDataProvider` is private (due to constraining construction/access to the instance getter `ForcingsEngineDataProvider::instance`), so without `new` it causes a compilation error -- maybe my design isn't idiomatic though

Reply:
Ok, this is fine to leave as-is, then. As noted, `make_unique` really just exists for symmetry with `make_shared`.